tests/contrib/gunicorn/test_gunicorn.py

```diff
@@ -52,79 +94,103 @@ def _gunicorn_settings_factory(
 bind = "{bind}"
 """.format(
-        post_worker_init=gunicorn_server_settings.post_worker_init,
+        post_fork=post_fork,
+        post_worker_init=post_worker_init,
         bind=gunicorn_server_settings.bind,
         num_workers=gunicorn_server_settings.num_workers,
         worker_class=gunicorn_server_settings.worker_class,
     )
+    return cfg
+
+
+@contextmanager
+def gunicorn_server(gunicorn_server_settings, tmp_path):
+    cfg_file = tmp_path / "gunicorn.conf.py"
+    cfg = build_config_file(gunicorn_server_settings)
     cfg_file.write_text(stringify(cfg))
     cmd = []
     if gunicorn_server_settings.use_ddtracerun:
         cmd = ["ddtrace-run"]
     cmd += ["gunicorn", "--config", str(cfg_file), str(gunicorn_server_settings.app_path)]
     print("Running %r with configuration file %s" % (" ".join(cmd), cfg))
-    p = subprocess.Popen(
+
+    server_process = subprocess.Popen(
         cmd,
         env=gunicorn_server_settings.env,
         cwd=gunicorn_server_settings.directory,
-        stdout=sys.stdout,
-        stderr=sys.stderr,
+        stderr=subprocess.PIPE,
         close_fds=True,
         preexec_fn=os.setsid,
     )
     try:
         client = Client("http://%s" % gunicorn_server_settings.bind)
-        # Wait for the server to start up
         try:
-            client.wait(max_tries=20, delay=0.5)
+            client.wait(max_tries=100, delay=0.1)
         except tenacity.RetryError:
-            # if proc.returncode is not None:
-            # process failed
             raise TimeoutError("Server failed to start, see stdout and stderr logs")
-
-        yield client
-
+        time.sleep(SERVICE_INTERVAL)
+        yield server_process, client
         try:
             client.get_ignored("/shutdown")
         except Exception:
             pass
     finally:
-        p.terminate()
-        p.wait()
+        server_process.terminate()
+        server_process.wait()
 
 
-def test_basic(gunicorn_server):
-    r = gunicorn_server.get("/")
-    assert r.status_code == 200
-    assert r.content == b"Hello, World!\n"
+SETTINGS_GEVENT_APPIMPORT_PATCH_POSTWORKERSERVICE = _gunicorn_settings_factory(
+    worker_class="gevent",
+    use_ddtracerun=False,
+    import_sitecustomize_in_app=True,
+    patch_gevent=True,
+    start_service_in_hook_named="post_worker_init",
+)
+SETTINGS_GEVENT_POSTWORKERIMPORT_PATCH_POSTWORKERSERVICE = _gunicorn_settings_factory(
+    worker_class="gevent",
+    use_ddtracerun=False,
+    import_sitecustomize_in_postworkerinit=True,
+    patch_gevent=True,
+    start_service_in_hook_named="post_worker_init",
+)
 
 
-@pytest.mark.snapshot(ignores=["meta.result_class"])
 @pytest.mark.parametrize(
-    "gunicorn_server_settings", [_gunicorn_settings_factory(app_path="tests.contrib.gunicorn.wsgi_mw_app:app")]
+    "gunicorn_server_settings",
+    [
+        SETTINGS_GEVENT_APPIMPORT_PATCH_POSTWORKERSERVICE,
+        SETTINGS_GEVENT_POSTWORKERIMPORT_PATCH_POSTWORKERSERVICE,
```

Comment:
Are we adding any scenarios that end in failure modes? Or are we just testing the two scenarios which don't result in failure modes? Should we include any expected failure test scenarios (not too many but at least one or two)? I worry once we have a "solution" for this gevent support these two scenarios alone won't make it obvious that the "solution" is indeed a solution.

Reply:
I was wondering this as well. I wrote a bunch of other tests that I then deleted to save time in CI. Which of those do you think are the most important to add back? https://github.com/DataDog/dd-trace-py/pull/4962/commits/564037ac3868b140abb181476b810d4790ba4025

Reply:
I think in an ideal scenario where we don't have to worry about flakiness of starting up a new gunicorn server for each test, it would be great to keep all of them. However in our case, `SETTINGS_GEVENT_DDTRACERUN_PATCH` might be the most important to add back just to illustrate that scenarios where we use `ddtrace-run` with gevent workers results in profiler issues.